src/models/StyledComponent.js

```diff
@@ -117,52 +169,72 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
     }
 
     componentWillUnmount() {
-      if (this.unsubscribe) {
-        this.unsubscribe()
+      const { componentStyle } = this.constructor
+      if (componentStyle.isStaticRules) {
+        return
       }
+
+      this.unsubscribeFromContext()
     }
 
-    render() {
-      const { innerRef } = this.props
+    makeClassName(): string {
+      // I know this is ugly, but I have a really surprigins JSPerf
+      // pen showing this is ~11x faster than running [...classes].filter(Boolean).join(' ')
+      // (and uses way less memory from the two array allocations in ^)
+      // and  ~ 36 % faster than always adding a whitespace and then calling .trim at the end
+      const { className: propsClassName } = this.props
+      const { className: attrsClassName } = this.attrs
       const { generatedClassName } = this.state
-      const { styledComponentId, target } = this.constructor
+      const { styledComponentId } = this.constructor
 
-      const isTargetTag = isTag(target)
+      let className = propsClassName || ''
+      if (styledComponentId) {
```

Comment:
You should be able to improve perf by a bit more by not casting any values here i.e. `className.length` => `className.length > 0` and so on.

Also you can assume that `generatedClassName` and `styledComponentId` will always be present.

Also, remove the branch that is conditionally adding spaces. We don't care if it's a bit malformed (trailing spaces are fine).

Then if you additionally change the remaining if-statements to include `else if` it'll be much cleaner.

Also, can you post the jsperf for future reference here :) The result will be definitely right, but I don't want people to ask for it.